### PR TITLE
Add Qdrant container

### LINKED ISSUE #992: [Feature]: Add Qdrant test container
### Problem

[Qdrant](https://qdrant.tech/) is a vector search engine/database that is becoming popular for search and recommendations use cases. Qdrant provides docker images but there isn't currently an official .NET test container. This issue proposes adding one.

### Solution

Open a PR with Qdrant test container from https://github.com/russcam/qdrant-dotnet-client/tree/main/tests/Qdrant.Grpc.Tests/Container

### Benefit

Provides a qdrant test container for folks that need it

### Alternatives

Folks who need a qdrant test container each define it themselves

### Would you like to help contributing this enhancement?

Yes

### PR DESCRIPTION
**Pull request by russcam**
Closes #992

## What does this PR do?

This commit adds a qdrant container to the list of supported Testcontainers.

The qdrant container allows configuration of:

- an API key to authenticate to Qdrant
- an x509 certificate used to secure communication to Qdrant with Transport Layer Security
- a custom configuration file. See https://qdrant.tech/documentation/guides/configuration/

## Why is it important?

Provides a qdrant test container for folks that need it

## Related issues

#992

## How to test this PR

Run tests in Testcontainers.Qdrant.Tests

### CODE REVIEW

**Comment by russcam** [comment]
This is ready to review @HofmeisterAn

File: src/Testcontainers.Qdrant/QdrantBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
+using System.IO;
+using System.Security.Cryptography.X509Certificates;
+using Org.BouncyCastle.OpenSsl;
+using Org.BouncyCastle.Security;
```
Comment (HofmeisterAn) [inline src/Testcontainers.Qdrant/QdrantBuilder.cs:4]: Please move the usings (other classes too) into the global `Usings.cs` file.

File: src/Testcontainers.Qdrant/QdrantBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+	private QdrantBuilder(QdrantConfiguration dockerResourceConfiguration) : base(dockerResourceConfiguration) =>
+		DockerResourceConfiguration = dockerResourceConfiguration;
+
+	/// <summary>
+	/// A path to a configuration file with which configure the instance.
+	/// </summary>
+	/// <param name="configurationFilePath">The path to the configuration file</param>
+	public QdrantBuilder WithConfigFile(string configurationFilePath) =>
+		Merge(DockerResourceConfiguration, new QdrantConfiguration(configurationFilePath: configurationFilePath))
```
Comment (HofmeisterAn) [inline src/Testcontainers.Qdrant/QdrantBuilder.cs:35]: If we do not require the `ConfigurationFilePath` property, along with the other configuration properties, at some time later, such as within the Qdrant container instance, then there is no need to keep and store the value. I think we can remove those properties from `QdrantConfiguration`

File: src/Testcontainers.Qdrant/QdrantBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+			.WithBindMount(configurationFilePath, QdrantLocalConfigurationFilePath);
```
Comment (HofmeisterAn) [inline src/Testcontainers.Qdrant/QdrantBuilder.cs:36]:
```suggestion
			.WithResourceMapping(configurationFilePath, QdrantLocalConfigurationFilePath);
```

File: src/Testcontainers.Qdrant/QdrantBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	public QdrantBuilder WithApiKey(string apiKey) =>
+		Merge(DockerResourceConfiguration, new QdrantConfiguration(apiKey: apiKey))
+			.WithEnvironment("QDRANT__SERVICE__API_KEY", apiKey);
+	
+	/// <summary>
+	/// A certificate to use to enable Transport Layer Security (TLS). The certificate must contain the
+	/// private key.
+	/// </summary>
+	/// <param name="certificate">A certificate containing a private key</param>
+	public QdrantBuilder WithCertificate(X509Certificate2 certificate)
```
Comment (HofmeisterAn) [inline src/Testcontainers.Qdrant/QdrantBuilder.cs:52]: Can we not expect the developer to provide valid `cert.pem` and `key.pem` files? This way, we can avoid dealing with all the certificate-related issues and Bouncy Castle.

File: tests/Testcontainers.Qdrant.Tests/Testcontainers.Qdrant.Tests.csproj [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    <PropertyGroup>
+        <TargetFrameworks>net6.0</TargetFrameworks>
+        <IsPackable>false</IsPackable>
+        <IsPublishable>false</IsPublishable>
+    </PropertyGroup>
+    <ItemGroup>
+        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.4.1"/>
+        <PackageReference Include="coverlet.collector" Version="3.2.0"/>
+        <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5"/>
+        <PackageReference Include="xunit" Version="2.4.2"/>
```
Comment (HofmeisterAn) [inline tests/Testcontainers.Qdrant.Tests/Testcontainers.Qdrant.Tests.csproj:11]:
```suggestion
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2"/>
        <PackageReference Include="coverlet.collector" Version="6.0.0"/>
        <PackageReference Include="xunit.runner.visualstudio" Version="2.5.0"/>
        <PackageReference Include="xunit" Version="2.5.0"/>
```

**Review by HofmeisterAn — COMMENTED** [review]
> This is ready to review @HofmeisterAn

Thank you for your pull request. I've already seen the PR but haven't had the time to thoroughly review it yet. Spare time is limited 😅.

Most of it looks good! I just have a few minor suggestions and improvements.

In general, we don't expose an API for every configuration (e.g., environment variable) if the developer can configure it themselves through the generic API (it doesn't require in-depth knowledge or isn't crucial for starting a basic container). However, I'm okay with the exposed APIs, but I would like to clean up some parts related to the TLS implementation if needed. I'll need to take a closer look at it (I will look into it by end of this week).

**Comment by HofmeisterAn** [comment]
This pull request has been inactive for some time. I will close it for now. If you wish to continue working on it, please feel free to reopen it.

**Comment by russcam** [comment]
@HofmeisterAn please can we reopen this PR; I've rebased from `develop` and updated following PR feedback. If it's easier, I can always open a new PR, but I don't have the ability to reopen this one

**Comment by HofmeisterAn** [comment]
> @HofmeisterAn please can we reopen this PR; I've rebased from `develop` and updated following PR feedback. If it's easier, I can always open a new PR, but I don't have the ability to reopen this one

I am sorry, I cannot reopen it either, it says:

> The qdrant-container branch was force-pushed or recreated.

You probably need to create a new PR - sorry.

**Comment by russcam** [comment]
Superseded by #1149

**Pull request closed without merging**